src/crypto/noise/mod.rs

```diff
@@ -709,62 +709,71 @@ pub async fn handshake<S: AsyncRead + AsyncWrite + Unpin>(
                 .map_err(ParseError::from)
                 .map_err(|err| {
                     tracing::error!(target: LOG_TARGET, ?err, "failed to decode remote identity message");
                     err
                 })?;
 
-            // send the final message which contains local peer id
-            let second_message = noise.second_message()?;
-            let _ = io.write(&second_message).await?;
-            io.flush().await?;
+                // send the final message which contains local peer id
+                let second_message = noise.second_message()?;
+                let _ = io.write(&second_message).await?;
+                io.flush().await?;
 
-            payload
-        }
-        Role::Listener => {
-            // read remote's first message
-            let _ = noise.read_handshake_message(&mut io).await?;
-
-            // send local peer id.
-            let second_message = noise.second_message()?;
-            let _ = io.write(&second_message).await?;
-            io.flush().await?;
-
-            // read remote's second message which contains their peer id
-            let message = noise.read_handshake_message(&mut io).await?;
-            // Decode the remote identity message.
-            handshake_schema::NoiseHandshakePayload::decode(message).map_err(ParseError::from)?
-        }
-    };
+                payload
+            }
+            Role::Listener => {
+                // read remote's first message
+                let _ = noise.read_handshake_message(&mut io).await?;
+
+                // send local peer id.
+                let second_message = noise.second_message()?;
+                let _ = io.write(&second_message).await?;
+                io.flush().await?;
+
+                // read remote's second message which contains their peer id
+                let message = noise.read_handshake_message(&mut io).await?;
+                // Decode the remote identity message.
+                handshake_schema::NoiseHandshakePayload::decode(message)
+                    .map_err(ParseError::from)?
+            }
+        };
 
-    let dh_remote_pubkey = noise.get_handshake_dh_remote_pubkey()?;
-    let peer = parse_and_verify_peer_id(payload, dh_remote_pubkey)?;
+        let dh_remote_pubkey = noise.get_handshake_dh_remote_pubkey()?;
+        let peer = parse_and_verify_peer_id(payload, dh_remote_pubkey)?;
 
-    Ok((
-        NoiseSocket::new(
-            io,
-            noise.into_transport()?,
-            max_read_ahead_factor,
-            max_write_buffer_size,
-        ),
-        peer,
-    ))
+        Ok((
+            NoiseSocket::new(
+                io,
+                noise.into_transport()?,
+                max_read_ahead_factor,
+                max_write_buffer_size,
+            ),
+            peer,
+        ))
+    };
+
+    match tokio::time::timeout(timeout, handle_handshake).await {
+        Err(_) => Err(NegotiationError::Timeout),
+        Ok(result) => result,
+    }
```

Comment:
Are you sure we won't end up in the inconsistent state if the future is canceled due to timeout? I.e., we should at least not reuse the `io` object if the future was interrupted?

Reply:
I think this is fine here since we consume the given io `mut io: S` and return a `NoiseSocket<S>` only on success. We have a similar pattern for `/noise` and `/yamux` negotiations 🙏